### ISSUE TITLE
MoE Q4 sorted-pairs path sums non-finite expert outputs: intermittent all-NaN logits on GB10 (sm_121)

### ISSUE THREAD

**Issue by AlexIannu** [antirez/ds4#1025]
Summary

On a DGX Spark (GB10, `sm_121`), DeepSeek V4 Flash with mixed q2-q4 quantization
degenerates intermittently: after the first token, **every** decode step produces
an all-NaN logit vector, `sample_argmax_unrolled8` never finds a maximum and
stays on index 0, which is `<|begin_of_sentence|>`. The visible result is a
stream of special tokens.

Measured over 45 independent restarts, **19 (42%) produced at least one ruined
generation**.

The proximate cause is that the **Q4 sorted-pairs MoE path is the only MoE path
in the engine that sums expert contributions without an `isfinite` check**. The
Q4 expert computation does produce canonical NaNs (`0x7fffffff`) during prefill;
`moe_mmq_sum_kernel`, `moe_mmq_swiglu_weighted_clamp_kernel` and the attention
softmax all discard non-finite values, this path propagates them into the
residual stream, then into the KV cache, and decode reads them back.

Environment

- **Dell Pro Max with GB10** (the Dell edition of the NVIDIA DGX Spark),
  `aarch64`, DGX OS 7.5.0 on Ubuntu 24.04.4, kernel 6.17.0-1031-nvidia.
- GPU **NVIDIA GB10 Blackwell**, compute capability **12.1 (`sm_121`)**, driver
  **580.173.02**, **CUDA 13.0** (V13.0.88). 128 GB unified LPDDR5X, no dedicated
  VRAM.
- ds4 at commit `b6af0adf8ca97c89145c9f9c15be70c9fd6c4507`, built with
  `make cuda-spark CUDA_ARCH=sm_121` (i.e. `-O3 --use_fast_math -gencode
  arch=compute_121a,code=sm_121a`).
- Model: `DeepSeek-V4-Flash-Layers37-42Q4KExperts-OtherExpertLayersIQ2XXSGateUp-Q2KDown-AProjQ8-SExpQ8-OutQ8-chat-v2-imatrix-fixed-0731.gguf`
  (90.89 GiB). Only layers 37-42 take `q4k_path`; the other expert layers are
  IQ2_XXS and go through the aligned artifacts path.
- Reproduced with `ds4-server` on `127.0.0.1`, `--ctx 8192`, no DSpark/MTP,
  temperature 0. Reproduced both with and without
  `--disable-exact-dsml-tool-replay` at the same rate (see below); that flag is
  unrelated.

How it shows up

Only the **first request after a restart** is likely to be hit, and at most one
request per process: over 20 restarts with two requests each, 9 restarts had the
first request ruined, 2 had the second one ruined, and **none had both**.

Adding a probe that counts non-finite terms right before the sum:

| restarts | non-finite terms in `down` | NaN decode steps | outcome |
|---|---|---|---|
| 3 of 8 | 0 | 0 | clean |
| 4 of 8 | 5 to 20+ | 79 of 80 | **ruined** |
| 1 of 8 | 3 | 0 | clean (see below) |

Every ruined run has non-finite values in the per-expert `down` buffer; no run
without them was ever ruined. 50 of 61 observed values were exactly
`0x7fffffff`, i.e. a **computed** NaN (`inf-inf` or `0*inf`), not garbage memory.
`compute-sanitizer --tool memcheck --kernel-name 'regex=moe'` went through model
load and the whole prefill with **zero** out-of-bounds accesses.

The one clean run that still produced NaNs is explained by the attention
softmax guard (`float p = isfinite(row[k]) ? expf(row[k] - max_s) : 0.0f;`): the
NaN landed on a key that then got silently ignored. So the 42% figure is a
**lower bound** on how often the model is affected.

It also corrupts tool calls, sometimes silently

Re-running the same probe with a tool-calling conversation instead of plain
prompts (2026-09-10, same unguarded build, 20 restarts) shows two distinct
shapes, both starting on the **first request after the restart**:

Visible — the DSML block collapses mid-emission:

```
<｜DSML｜tool_calls>
<｜DSML｜invoke name="leggi_misura">
<｜DSML｜parameter name="nome"<｜begin▁of▁sentence｜><｜begin▁of▁sentence｜>…
```

logged by the server as
`TOOLS invalid tool call returned as assistant text finish=length [saw_start=1 saw_end=0]`.

**Silent, and worse** — a structurally valid tool call with a corrupted function
name. This one returned `tool_calls=1`, `finish_reason=tool_calls`, no special
tokens and no repetition, i.e. it passes every marker-based or
repetition-based check, but the function name was `leggi_m FOX` instead of
`leggi_misura`. An agent harness would dispatch a call to a tool that does not
exist and surface an unrelated error downstream.

This matters for anyone using the engine agentically: **a marker-only degeneracy
check under-reports the defect**, and the corruption can reach the tool
dispatcher rather than the user's screen.

Not the DSML tool replay map

Worth stating because the symptom looks like it. `--disable-exact-dsml-tool-replay`
makes **no difference** to the failure rate: 1/10 restarts ruined with the flag,
1/10 without it, on the same unguarded build (Fisher exact p = 1.0). In both
ruined runs the failure is on **turn 0**, where the trace shows

```
tool_replay: mem=0 disk=0 canonical=0 missing_ids=0
```

i.e. no tool history exists yet and `tool_memory_*` returns immediately, so the
replay map is not in play at all when the corruption appears.

## Where the guard is missing

`moe_sum_kernel` (`ds4_cuda.cu:23896` at this commit) sums unconditionally:

```c
for (uint32_t e = 0; e < n_expert; e++) acc += down[((uint64_t)tok * n_expert + e) * out_dim + row];
```

while `moe_mmq_sum_kernel` (`ds4_cuda.cu:1183`), used by the MXFP4, MMQ and
aligned IQ2 paths, has had the guard all along:

```c
if (!guard_nonfinite || isfinite(v)) acc += v;
```

The decode-side direct-sum kernels have the same gap, and this one bites *during*
generation rather than at the first token:

- `moe_down_sum6_qwarp32_kernel` (`:22018`)
- `moe_down_sum3_qwarp32_kernel` (`:22216`)
- `moe_down_q4K_sum6_qwarp32_kernel` (`:22248`)
- `moe_down_q4K_sum3_qwarp32_kernel` (`:22490`)

all doing `if (lane == 0) total += acc;`.

Note that the clamp in the tile kernels does not help: a NaN fails every
comparison, so `if (gate[p] > clamp) gate[p] = clamp;` leaves it untouched.

Fix that works here

Five one-line changes, applying the convention the engine already uses
elsewhere — treat an expert whose contribution is not finite as contributing
zero:

```c
/* moe_sum_kernel */
const float v = down[((uint64_t)tok * n_expert + e) * out_dim + row];
if (isfinite(v)) acc += v;

/* the four direct-sum kernels */
if (lane == 0 && isfinite(acc)) total += acc;
```

Measured on the machine above, 20 independent restarts per arm, ctx 8192:

| | before | after |
|---|---|---|
| restarts with a ruined generation | 11 / 20 | **0 / 20** |
| `7 * 8` answered exactly `56` | 18 / 20 | **20 / 20** |
| prefill (mean from ctx 8192 up) | 800.1 t/s | 816.8 t/s |
| decode (first chunk, clean requests) | 13.08 t/s | 12.96 t/s |

Both throughput figures are within run-to-run noise: the guard is free. The
probe still fires in **12 of 20 restarts** afterwards, so the NaNs keep being
produced at the same rate, they just no longer reach the output.

Guarding only `moe_sum_kernel` was not enough: it left 2 ruined runs out of 20,
with the avalanche starting mid-sentence instead of at the first token, which is
what pointed at the decode-side sums.

Held up under wider conditions afterwards (2026-09-10): **44 further restarts,
492 tool-calling turns, 0 ruined**, spanning ctx 8k / 100k / 1M, with and
without DSpark (`--dspark --mtp-exact-sampling`), and tool results from 20 bytes
to 9 KB. Every tool call had the correct function name and argument. 95%
confidence bound on the per-restart failure rate afterwards: **0 - 8%** (0 -
0.77% per request).

Workaround without patching

`DS4_CUDA_MOE_NO_Q4_SORTED=1` also gives 0/20, because it makes layers 37-42
fall back to a guarded path — but it costs **3.11x on prefill** (688 -> 221 t/s
in our measurement), so it is not a good trade.

What I could not determine

Which operation actually creates the NaN.** It is born in the Q4 expert
product during prefill and it is a canonical NaN, so something upstream
overflows to infinity first. Candidates I did not manage to separate: the
`dev_dot_q4_K_q8_K_block*` accumulation, the q8_K quantization of `mid`, or an
overflow made more likely by `--use_fast_math` (`-ftz`, approximate intrinsics)
which is in the project's build flags. A guard is a mitigation; the arithmetic
question is the real bug and I am happy to run further experiments on this
hardware if that helps.

Two hypotheses I did test and **ruled out**, in case they save someone time:
uninitialized reads of the sorted-pairs scratch (zeroing the entire scratch
leaves the failure rate unchanged at 40%), and out-of-bounds writes (memcheck
clean across prefill).
